### DISCUSSION TITLE
Overwriting smooth fonts

### DISCUSSION THREAD

**Discussion by Phrazemaster** [Bodmer/TFT_eSPI#1826] [Q&A - General] [answered]
Hi Mr. Bodmer, what an absolutely fantastic library you've made. You're a genius!! Thank-you so much! You must never sleep.

I am having some difficulty with some large numbers and text. The text padding works horizontally, but some of my letters and numbers need a padding on top. Code is below and a screenshot so you can see. The "lines" above the number are from previous numbers and text that did not get overwritten when values changed.

Any thoughts? I've tried using your new flicker free tft.print function with the setCursor(x,y,true) which works great, no extra detritus at the tops of the letters/numbers, but it seems the tft.print function does not support text datum, and I need this to be centered. Adding a rectangle behind only makes terrible flicker. I've also tried using sprites - which work great, except for a bizarre case where if I use green text color, the numbers and text are pixelated and not anti-aliased...very odd because I have 4 sensor values I'm printing in different colors and the other 3 colors print smoothly - same code, just different color. I don't want to hijack my own thread, so I'll leave that for another day and focus on the issue with the drawString results.

So I'm hoping there might be a way to get the drawstring to work as in the example below.

Thanks for any help.

Code:
tft.setTextDatum(TC_DATUM);
tft.setTextPadding(30); 
tft.loadFont(ArialBold126);
tft.setTextColor(TFT_YELLOW, TFT_BLACK, true);                                         
tft.drawString("     " + String(SensorVal + "     "), xC,Ynum );

Mike
![tops-of-letters-and-numbers-not-overwritten2](https://user-images.githubusercontent.com/17349138/167984634-e850fdff-77d7-4e6c-b4fc-935560246c8c.jpg)

**Comment by Phrazemaster** [comment]
I should add, a previous version of the library did not exhibit this behavior. Thanks.

**Comment by Bodmer** [comment]
The library examples do not exhibit the problem so this may be specific to the font you have created. To identify the root cause I need a simple sketch that demonstrates the problem, this must include the font.

Incidentally the purpose of setTextPadding() is to avoid the need to pad with spaces. Also using drawNumber with an appropriately set padding width will avoid the need to convert to a string.

**Comment by Phrazemaster** [comment]
Thank you, I can appreciate that. I will see what I can do to create a simple version for you to test. I made the font using your amazing Processing sketch, and it's Arial Bold at 126 size.

I also found some *VERY VERY* strange results when using the text padding - to the point I had to stop using it. It broke completely unrelated things, and it was only sheer intuition that led me to try disabling it and then normal functionality was restored. I realize this helps you 0%, I'm sorry. But there's definitely some issues with it.

Working code with a previous version of your library was what broke after I tried using the text padding.

I'll see if I can make something for you to test regarding the numbers. I made a workaround with a small rectangle over the tops of the numbers that isn't noticeable but removes the offending leftovers. Using a previous value and painting it underneath only led to terrible flicker.

Again, I can only humbly say thank-you. I'm not sure people realize what a monumental triumph this library is, and your work in general. Hats off!

Mike

**Comment by Bodmer** [comment]
I created the attached in an attempt to re-create the problem, however it works fine.

Run it and see if you have a problem. The font only has 0-9 in it.

[Issue_1826.zip](https://github.com/Bodmer/TFT_eSPI/files/8694094/Issue_1826.zip)

If unrelated things get broken then this suggests a memory consumption problem, or running off then end of an array elsewhere in your sketch.

**Reply by Phrazemaster** [reply]
Thank you so much. I don't know how you find the time to provide such stellar support.

Your sketch works fine. But mine was working flawlessly before updating to the most recent TFT lib, and the only thing I changed was the lib. However, again, this doesn't help you.

It definitely appears that some of the sprites poke out of the tops of their normal boundaries - I've gotten some very odd breakage of surrounding boxes near a sprite. Changing back to printing with tft.drawString and using your new updated tft.setTextColor(x,y,true) seems to work instead.

Again - no change to my working code but the lib update.

I have so many versions now of this commercial sketch and I'm afraid I can't find the one with the problem now, since I did the above workarounds.

What amazes me is how incredible your library and font smoothing really are. Nothing comes close.

I really appreciate your time. Thank-you.

**Comment by Bodmer** [comment, marked as answer]
Glad to here you have a solution. As far as I can tell TFT_eSPI is working correctly.

The "Alerts" section is for identifiable problems and solutions  so to avoid misunderstanding I have moved the discussion and renamed it.

**Reply by Phrazemaster** [reply]
I understand a vague description is useless, sorry. But there is some kind of problem. Maybe it is with my sketch but I've been careful. At one point, the text from a previous tft.print appeared inside of sprites! And they displaced the text that they were set to show. Only when I stopped using sprites for that portion of the program did it go away.

I remain amazed at your achievements. Thank-you.

**Comment by Phrazemaster** [comment]
Hi Mr. Bodmer, I am posting a picture here of the results of using sprites with smooth font NotoSansBold56 generated with Processing. 

As you can see, the letter "G", and the third letter "M" have some pixelated debris around them, and the middle letter "M"does not. Strangely, the ones with the pixelated debris only have it partially; much of the letter is rendered with smooth antialiasing.  I will post a snippet of code but it's too complex for me to fully isolate.

Sorry it appears to only have recognized part of the code for formatting; I'm not sure how to fix this.

while(blnEditingCCName) {
          analog[0].update();   
          if(analog[0].hasChanged()) {   
            Data[0] = analog[0].getValue()>>3;           
            if (DataLag[0] != Data[0]){
  
              Fader1Pos = CharPositionInverted[round(Data[0]/4 + .5) - 1];//32 letters/symbols, 0-based
  
              spr1.deleteSprite();
              spr1.createSprite(50,50);
              spr1.fillSprite(TFT_TRANSPARENT);
              spr1.loadFont(NotoSansBold56);
              spr1.setTextDatum(TC_DATUM);
              spr1.setTextColor(qec, bg, true);
              spr1.setTextDatum(TC_DATUM);
              spr1.setTextPadding(50);
    
              //First CC Name Letter       
              spr1.drawString(NameChars[Fader1Pos], 25,0);
              spr1.pushSprite(XNameStart - 2, UnderLineDown - 51, TFT_TRANSPARENT);
              spr1.setTextPadding(0);
              spr1.deleteSprite();                         
              
            }// if (DataLag[0] != Data[0])
          }//if(analog[0].hasChanged()

}//end while

Do you have any suggestions as to what could possibly cause this? I've had to work around this several times and it's not always easy. Your sprites are terrific but they seem to cause this, and I notice it often with a TFT_GREEN color and not much with other colors. If I just use tft.drawstring's it does not have the issue - but then there's another problem with the tft rendering a huge box below the text that causes flicker with other elements. So the sprites work perfectly other than this pixelation.
![IMG_9218](https://user-images.githubusercontent.com/17349138/168512176-21f43fab-0b7a-4bb4-a253-ff26423c8420.jpeg)


Thanks for your thoughts.

Mike

**Reply by Bodmer** [reply]
The latest 2.4.61 has been released with some bug fixes, so would be worth trying.

**Comment by Phrazemaster** [comment]
No worries! And thank- you so much for letting me know.

In my own programming I've come to have immense respect for the complexity that goes into projects like yours. I'm amazed more things didn't break; the new updates are terrific!

I did have another little wonder - sometimes when compiling, I get messages about unused variables coming from the TFT library. The messages don't show up every time either. Any thoughts on that?

Thanks again. I hope somehow you get paid for all of this monumental effort.

Mike

![Screen Shot 2022-05-16 at 11 50 57 AM](https://user-images.githubusercontent.com/17349138/168662134-8ea49fd9-3b87-4e89-abc5-dd6af1f9a3e6.jpg)

**Reply by Bodmer** [reply]
The warnings will occur if you have disabled some fonts in the setup. The warnings will come and go if you are using different setups with different fonts enabled/disabled.
